### PR TITLE
feature: Create update product screen

### PR DESCRIPTION
**Pull request by halfangel37**
## Change Description :wave:

- v-model remove
- services refactor
- upload image refactor

## Demo 📺
https://www.loom.com/share/6507e271d9bc46f7889650b163d8e98
New update 19/7:
Refactor Update User Profile and Create Product:
https://www.loom.com/share/3f6fe97caea6436a947d3724e9fceb87
New Feature Update product screen:
https://www.loom.com/share/3d27e6baedb047a3920c501cd0cce555

### CODE REVIEW

File: src/shared/form-validators/index.js [hunk 1]
```diff
@@ -11,7 +11,9 @@ const email = (message) => (value) => {
 const required = (message) => (value) => {
   return (value && value.length > 0) || message;
 };
-
+const requiredNumber = (message) => (value) => {
+  return (value && value!==0) || message;
+};
```
Comment (luantc13) [inline src/shared/form-validators/index.js:16]: `0` is also a number. I think you should have a validator called "requiredPositiveNumber`.

File: src/product/views/UpdateProduct.vue [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      productId: this.$route.params.productId,
+    });
+  },
+  methods: {
+    updateProduct(product) {
+      this.isPending = true;
+      this.$store
+        .dispatch(
+          "product/updateProduct",
+
```
Comment (luantc13) [inline src/product/views/UpdateProduct.vue:39]: A redundant empty line.

File: src/product/views/UpdateProduct.vue [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+          {
+            companyId: this.$route.params.companyId,
+            productId: this.$route.params.productId,
+
```
Comment (luantc13) [inline src/product/views/UpdateProduct.vue:43]: A redundant empty line.

File: src/product/store/index.js [hunk 4]
```diff
@@ -15,5 +22,27 @@ export default {
           Vue.$toast.error(error.response.data.errors);
         });
     },
+    getProductDetail({ commit }, { companyId, productId }) {
+      return httpClient
+        .get(`/companies/${companyId}/products/${productId}`)
+        .then((response) => {
+          commit("SET_PRODUCT", response.data);
+        });
```
Comment (luantc13) [inline src/product/store/index.js:30]: I don't think we should use `httpClient` directly.

File: src/product/components/UpdateProductForm.vue [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+  computed: {
+    currencies() {
+      return currencyCodes.map((item) => item.cc);
+    },
+    currenciesArray() {
+      const mapCurrencies = this.cloneProduct.productPrices.map(
+        (item) => item.currency
+      );
+      return mapCurrencies;
+    },
```
Comment (luantc13) [inline src/product/components/UpdateProductForm.vue:153]: `currencies` and `currenciesArray` make me confused, pls rename `currenciesArray` to `productCurrencies`.

File: src/product/components/UpdateProductForm.vue [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+export default {
+  props: {
+    isPending: Boolean,
+    product: Object,
+  },
+  data() {
+    return {
+      nameRules: [validators.required("Product name is required")],
+      priceRules: [validators.requiredNumber("Product price is required")],
+
```
Comment (luantc13) [inline src/product/components/UpdateProductForm.vue:104]: A redundant empty line.

**Review by luantc13 — CHANGES_REQUESTED** [review]
Nice work!
Having some small comments, happy to merge when they are resolved.

**Comment by truonghungit** [comment]
the first time does not show any things 
 https://www.loom.com/share/e8ec85eadc8f498b9f034a4c6f5eb050

File: src/product/views/CreateProduct.vue [hunk 7]
```diff
@@ -22,15 +22,22 @@ export default {
     isPending: false,
   }),
   methods: {
-    createProduct(product) {
+    createProduct(product, File) {
```
Comment (luantc13) [inline src/product/views/CreateProduct.vue:25]: I don't think `File` is a good name, pls rename it to `formData`.

File: src/http/products.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+};
+const createProduct = ({ companyId, product }) => {
+  return httpClient.post(`${COMPANY}/${companyId}${PRODUCTS}`, product);
+};
+const createProductImage = ({ companyId, productId, File }) => {
+  return httpClient.post(
+    `${COMPANY}/${companyId}${PRODUCTS}/${productId}${IMAGE}`,
+    File
+  );
+};
```
[the remark this answers is not in the document]
Comment (luantc13) [reply src/http/products.js:29]: I don't think `File` is a good name, pls rename it to `formData`.

File: src/product/components/UpdateProductForm.vue [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        class="avatar"
+        :src="
+          imageUrl +
+          this.user.id +
+          '/companies/' +
+          this.$route.params.companyId +
+          '/products/' +
+          this.product.id +
+          '/' +
+          this.product.imageName
```
Comment (luantc13) [inline src/product/components/UpdateProductForm.vue:21]: This is hard to read, may we use another approach?

File: src/product/store/index.js [hunk 10]
```diff
@@ -1,19 +1,75 @@
...
+  },
   actions: {
-    createProduct(_, { product, companyId }) {
-      return httpClient
-        .post(`/companies/${companyId}/products`, product)
-        .then(() => {
+    createProduct({ commit }, { companyId, product }) {
+      return createProduct({ companyId, product })
+        .then((response) => {
+          commit("GET_PRODUCT_ID", response.data.id);
```
Comment (luantc13) [inline src/product/store/index.js:29]: `GET_PRODUCT_ID`? Why do we need this mutation?

File: src/profile/components/UserProfileForm.vue [hunk 11]
```diff
@@ -7,7 +7,7 @@
       ></v-img>
       <v-file-input
         @change="onFileChange"
-        v-model="user.imageName"
+        v-model="cloneUser.imageName"
```
Comment (luantc13) [inline src/profile/components/UserProfileForm.vue:10]: They are have the different data type.

File: src/profile/components/UserProfileForm.vue [hunk 12]
```diff
@@ -62,20 +62,33 @@ export default {
...
-  data: () => ({
-    imageUrl: imageBaseUrl,
-    nameRules: [validators.required("Name is required")],
-  }),
-
+  data() {
+    return {
+      imageUrl: imageBaseUrl,
+      nameRules: [validators.required("Name is required")],
+      cloneUser: "",
```
Comment (luantc13) [inline src/profile/components/UserProfileForm.vue:69]: `cloneUser` is an object, not a string.

**Review by luantc13 — CHANGES_REQUESTED** [review]
Nice work!
Having some comments, happy to merge when they are resolved.

File: src/http/endpoints.js [hunk 13]
```diff
@@ -9,4 +9,6 @@ export const FORGOT_PASSWORD_ENDPOINT = AUTH + "/forgot-password";
 export const RESET_PASSWORD_ENDPOINT = AUTH + "/reset-password";
 export const CHANGE_PASSWORD_ENDPOINT = ACCOUNT + "/change-password";
 export const COMPANY_ENDPOINT = "/companies";
-export const DELETE_COMPANY_ENDPOINT = "/companies";
\ No newline at end of file
+export const PRODUCTS = "/products";
+export const IMAGE = "/image";
+export const DELETE_COMPANY_ENDPOINT = "/companies";
```
[the remark this answers is not in the document]
Comment (halfangel37) [reply src/http/endpoints.js:14]: export const DELETE_COMPANY_ENDPOINT = "/companies";
its already in master branch, I pull master and it was there.

File: src/product/store/index.js [hunk 14]
```diff
@@ -1,19 +1,66 @@
...
   actions: {
-    createProduct(_, { product, companyId }) {
-      return httpClient
-        .post(`/companies/${companyId}/products`, product)
-        .then(() => {
+    createProduct(_, { companyId, product, formData }) {
+      return createProduct({ companyId, product })
+        .then((response) => {
           Vue.$toast.success("Product have been created.");
+          let productId = response.data.id;
```
Comment (luantc13) [inline src/product/store/index.js:26]: I think we should replace `let` with `const`, I see that `productId` is only assigned a value once.

File: src/product/store/index.js [hunk 15]
```diff
@@ -1,19 +1,66 @@
...
+  updateProduct,
+  uploadProductImage,
+} from "@/http/products.js";
 import Vue from "vue";
 export default {
   namespaced: true,
-  state: {},
-  mutations: {},
+  state: {
+    product: { imageName: "" },
```
Comment (luantc13) [inline src/product/store/index.js:11]: Pls explain more about the default value of `imageName`.

**Review by luantc13 — APPROVED** [review]
Great work!

File: src/product/store/index.js (hunk 15 above)
Comment (halfangel37) [reply src/product/store/index.js:11]: I do it before to make the image change reactively, but not I check it again and see that I dont need it , i just need the 
image
to make it reactively, so I will delete it.

**Review by truonghungit — APPROVED** [review]

**Comment by truonghungit** [comment]
@halfangel37 please wait for [PR #40](https://github.com/ezteksoftware/painite/pull/40)

**Pull request merged**